core/src/main/scala/spire/optional/Perm.scala

```diff
@@ -0,0 +1,79 @@
+package spire
+package optional
+
+import scala.collection.compat._
+import spire.scalacompat.SeqLike
+
+import cats.kernel.Eq
+import spire.algebra.{Action, Group}
+import spire.algebra.partial.PartialAction
+import spire.syntax.cfor._
+import spire.syntax.group._
+import spire.util._
+
+/**
+ * Represents a permutation encoded as a map from preimages to images, including
+ * only pairs that are moved by the permutation (so the identity is Map.empty).
+ */
+class Perm private(val mapping: Map[Int, Int]) extends (Int => Int) {
```

Comment:
Is a permutation a function `Int => Int`, or a `PartialFunction`?

Reply:
It's total. A value not in the Map is mapped to itself.

Reply:
I see how this represents the group of permutations over all integers (or all Int).
For people using it only w.r.t. the subset that is actually permuted, it allows a composition of permutations whose domains and ranges don't align. That is probably fine, and checking it would subtract from its current elegance, but bikeshedding mathematical semantics is sort of spire's brand so... :smile: 
Maybe also just a matter of documentation.

Reply:
Yeah, I'd like restricted perms too, but that would require more overhead (include identity mappings?), and depart from the previous api. For what it's worth, AlaSc and GAP both use the same semantics. Perhaps one could use a type tag like `Perm @@ S3` after checking that the image is contained by 0 until 3.

I'll make the doc a bit more explicit about image vs domain.